units/event_metajob.go

```diff
@@ -159,28 +162,51 @@ func (j *eventMetaJob) dispatchLoop(ctx context.Context) error {
 	startTime := time.Now()
 	logger := event.NewDBEventLogger(event.AllLogCollection)
 	catcher := grip.NewSimpleCatcher()
-	notifications := make([][]notification.Notification, len(j.events))
 
-	var err error
-	for i := range j.events {
-		notifications[i], err = tryProcessOneEvent(&j.events[i])
-		catcher.Add(err)
-		if err = notification.InsertMany(notifications[i]...); err != nil {
-			grip.Error(message.WrapError(err, message.Fields{
-				"job_id":        j.ID(),
-				"job":           eventMetaJobName,
-				"source":        "events-processing",
-				"notifications": notifications[i],
-				"message":       "can't insert notifications",
-			}))
-			catcher.Add(err)
+	input := make(chan event.EventLogEntry, len(j.events))
+	for _, e := range j.events {
+		select {
+		case input <- e:
+			continue
+		case <-ctx.Done():
+			return errors.New("operation aborted")
 		}
-	}
 
-	for idx := range notifications {
-		catcher.Add(j.dispatch(notifications[idx]))
-		catcher.Add(logger.MarkProcessed(&j.events[idx]))
 	}
+	close(input)
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < runtime.NumCPU(); i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer recovery.LogStackTraceAndContinue("problem during notification dispatch")
+
+			for e := range input {
+				n, err := tryProcessOneEvent(&e)
+				catcher.Add(err)
+				if err != nil {
```

Comment:
This is different from the previous implementation. That implementation calls `InsertMany` even if there is an error. One problem with this implementation is that if `tryProcessOneEvent` or `notification.InsertMany` fail, `logger.MarkProcessed` will not run, potentially causing us to reprocess events with some data problem on every run of the metajob.

Reply:
We should probably move MarkProcessed up, but I think the rest of this loop is fine. We shouldn't ever return any notifications if there are errors (it would be a bad pattern to follow anyway)

Reply:
I'm down to move mark processed up, but I don't actually think this is very different what was happening before in the two loops. 

I think another, future improvement of this job would be to create different jobs for each even processor, so that there aren't threads/loops here, but I think this is probably a reasonable step for now.